base/ca/src/org/dogtagpki/server/ca/rest/CertService.java

```diff
@@ -209,12 +217,30 @@ public Response revokeCert(CertId id, CertRevokeRequest request, boolean caCert)
             if (clientCert != null) {
                 clientSerialNumber = clientCert.getSerialNumber();
                 clientSubjectDN = clientCert.getSubjectDN().toString();
-                clientRecord = processor.getCertificateRecord(clientSerialNumber);
 
-                // Verify client cert is not revoked.
-                // TODO: This should be checked during authentication.
-                if (clientRecord.getStatus().equals(ICertRecord.STATUS_REVOKED)) {
-                    throw new UnauthorizedException(CMS.getLogMessage("CMSGW_UNAUTHORIZED"));
+                String clientIssuerDN = clientCert.getIssuerDN().toString();
+                X500Name x500issuerDN = new X500Name(clientIssuerDN);
+                /*
+                 * internal revocation check only to be conducted for certs
+                 * issued by this CA
+                 * For client certs issued by external CAs, TLS mutual auth
+                 * would have completed the authenticaiton/verification if
+                 * OCSP was enabled;
+                 * Furthermore, prior to the actual revocation, client cert
+                 * is mapped against the agent group database for proper
+                 * privilege regardless of the issuer.
+                 */
+                if (x500issuerDN.equals(caX500DN)) {
+                    CMS.debug("CertService.revokeCert: client cert issued by this CA");
+                    clientRecord = processor.getCertificateRecord(clientSerialNumber);
+
+                    // Verify client cert is not revoked.
+                    // TODO: This should be checked during authentication.
+                    if (clientRecord.getStatus().equals(ICertRecord.STATUS_REVOKED)) {
+                        throw new UnauthorizedException(CMS.getLogMessage("CMSGW_UNAUTHORIZED"));
+                    }
+                } else {
+                    CMS.debug("CertService.revokeCert: client cert not issued by this CA");
```

Comment:
Does this mean in case the OCSP was not enabled or not working properly the user will be allowed to proceed?

To my understanding the server needs to search the external cert in the user subtree to map it to the right user (i.e. the agent) and that user needs to be a member of Certificate Manager Agents Group before we allow it to proceed. Is that operation already done by `PKIRealm` or is it something that needs to be added here?

Reply:
the group membership check (with cert mapping) is done after the area where the code change happens.  And yes, if ocsp isn't enabled then the cert status check would be bypassed. Perhaps we could add a flag in CS.cfg to require explicitly enabling of this?  Like, maybe ca.allowExternalCAAgentCert=true ?

Reply:
I think the membership checking in line 248 is used to see that if nonces are enabled on the system and the user is not a member of `Subsystem Group` then the provided nonce will be validated. But if the nonces are not enabled, or if the user is a member of `Subsystem Group`, they can bypass nonce validation. So this code doesn't check whether the user is a member of `Certificate Manager Agents Group` as described in the bug.

Reply:
Ah never mind, I see another group validation in line 263.

Reply:
I was actually referring to the combination of principal.hasRole() call and processor.validateCertificateToRevoke() call which together makes sure that the cert is either revoked by the owner of the cert or an agent.
Would this do?

Reply:
Actually, I think line 263 means that agents can revoke any certs, but others can only revoke their own cert. The access to `revokeCert()` is controlled by this ACL:
```
resourceACLS: certServer.ca.certs:execute:allow (execute) group="Certificate Manager Agents":Agents may execute cert operations
```

Reply:
yes.  I thought that's what I said. ;-)  the hasRole() makes sure that the subjectDN turns to null so that the next call into processor.validateCertificateToRevoke() will allow the agent to revoke any cert and the matching user gets to revoke his/her own cert.
anyways, I'm about to submit the updated patch with the allowExtCASignedAgentCerts idea.  Thanks.

Reply:
OK, I think actually there are 3 possible revocation checks:
* If `enableOCSP` is enabled in `server.xml`, the external client cert will be checked against the external OCSP.
* If internal revocation checking is enabled, the client cert revocation status will be checked again in [CertUserDBAuthentication.authenticate()](https://github.com/dogtagpki/pki/blob/DOGTAG_10_5_BRANCH/base/server/cmscore/src/com/netscape/cmscore/authentication/CertUserDBAuthentication.java#L156).
* Regardless of the above checks, `CertService.revokeCert()` will check the revocation status again.
 
The second check actually has a problem since it'll try to find the cert record based on the client cert's serial number, which may not exist if the cert was issued by an external CA, but it's only a problem if the internal revocation checking is enabled. Maybe it should do an OCSP check if the cert was external.

The patch is correct that it will skip the third check if the client cert was external so I'll ACK this PR. No need to add another parameter. Ideally the second check should be fixed too, but maybe we can do that another time.

Reply:
oh I already added the ca.allowExtCASignedAgentCerts param for CS.cfg.  The purpose is to make admin aware that that's what they are getting.  Hope this is not going to revoke the ACK? ;-)

Reply:
I don't think the param is needed since if they don't want an external cert in the first place, they probably wouldn't assign it to the user. No big deal though.

Reply:
ok so basically by default, nothing changes lat least. ;-)  Pushing now.

Reply:
There's actually one more possible issue. In line 256 it's comparing the serial number of the cert to be revoked and the serial number of the client cert, which maybe from an external CA, so the serial numbers may be the same but they are different certs. I'm not sure the implication of this, maybe we can deal with it if it becomes a problem.